vc/vc.go

```diff
@@ -0,0 +1,198 @@
+package vc
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/tbd54566975/web5-go/dids/did"
+	"github.com/tbd54566975/web5-go/jwt"
+)
+
+// these constants are defined in the W3C Verifiable Credential Data Model specification for:
+//   - [Context]
+//   - [Type]
+//
+// [Context]: https://www.w3.org/TR/vc-data-model/#contexts
+// [Type]: https://www.w3.org/TR/vc-data-model/#dfn-type
+const (
+	BaseContext = "https://www.w3.org/2018/credentials/v1"
+	BaseType    = "VerifiableCredential"
+)
+
+// DataModel represents the W3C Verifiable Credential Data Model defined [here]
+//
+// [here]: https://www.w3.org/TR/vc-data-model/
+type DataModel[T CredentialSubject] struct {
+	Context           []string `json:"@context"`                 // https://www.w3.org/TR/vc-data-model/#contexts
+	Type              []string `json:"type"`                     // https://www.w3.org/TR/vc-data-model/#dfn-type
+	Issuer            string   `json:"issuer"`                   // https://www.w3.org/TR/vc-data-model/#issuer
+	CredentialSubject T        `json:"credentialSubject"`        // https://www.w3.org/TR/vc-data-model/#credential-subject
+	ID                string   `json:"id,omitempty"`             // https://www.w3.org/TR/vc-data-model/#identifiers
+	IssuanceDate      string   `json:"issuanceDate"`             // https://www.w3.org/TR/vc-data-model/#issuance-date
+	ExpirationDate    string   `json:"expirationDate,omitempty"` // https://www.w3.org/TR/vc-data-model/#expiration
+}
+
+// CredentialSubject is implemented by any type that can be used as the CredentialSubject
+// of a Verifiable Credential.
+//
+// # Note
+//
+// The VC Data Model specification states that id is not a required field for [CredentialSubject]. However,
+// we've chosen to require it in order to necessitate that all credential's include a subject as we were unable
+// to find a use case where a credential would not be issued to a single subject. Further, the spec states that
+// [vc-jwt] requires the sub be set to the id of the CredentialSubject which becomes difficult to assert while
+// also providing the ability to leverage strongly typed claims.
+//
+// [CredentialSubject]: https://www.w3.org/TR/vc-data-model/#credential-subject
+// [vc-jwt]: https://www.w3.org/TR/vc-data-model/#json-web-token
+type CredentialSubject interface {
+	GetID() string
+	SetID(id string)
+}
+
+// Claims is a type alias for a map[string]any that can be used to represent the claims of a Verifiable Credential
+// when the structure of the claims is not known at compile time.
+type Claims map[string]any
```

Comment:
Maybe this should be called `GenericClaims` to convey its intent?

Reply:
good point! issue created: https://github.com/TBD54566975/web5-go/issues/80